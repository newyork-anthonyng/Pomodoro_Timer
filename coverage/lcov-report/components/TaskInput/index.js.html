<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for components/TaskInput/index.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> / <a href="index.html">components/TaskInput</a> index.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>8/8</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">50% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>1/2</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>3/3</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>8/8</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React, { PureComponent } from 'react';
import T from 'prop-types';
import noop from '../../utility/noop';
&nbsp;
class TaskInput extends PureComponent {
  constructor() {
    super();
&nbsp;
    this.handleKeyPress = this.handleKeyPress.bind(this);
  }
&nbsp;
  handleKeyPress(e) {
    <span class="missing-if-branch" title="else path not taken" >E</span>if (e.key === 'Enter') {
      this.props.onSubmit();
    }
  }
&nbsp;
  render() {
    const { value, onChange, onSubmit } = this.props;
&nbsp;
    return (
      &lt;div&gt;
        &lt;textarea
          value={value}
          onChange={onChange}
          onKeyPress={this.handleKeyPress}
        /&gt;
        &lt;button onClick={onSubmit}&gt;âœ“&lt;/button&gt;
      &lt;/div&gt;
    );
  }
}
&nbsp;
TaskInput.defaultProps = {
  onSubmit: noop,
  onChange: noop,
};
&nbsp;
TaskInput.propTypes = {
  value: T.string.isRequired,
  onSubmit: T.func.isRequired,
  onChange: T.func.isRequired,
};
&nbsp;
export default TaskInput;
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Fri May 05 2017 22:18:16 GMT-0700 (PDT)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
