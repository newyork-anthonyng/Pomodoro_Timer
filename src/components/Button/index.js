import React from 'react';
import T from 'prop-types';
import noop from '../../utility/noop';

const Button = ({ text, onClick, style }) => (
  <button style={style} onClick={onClick}>{text}</button>
);

Button.defaultProps = {
  onClick: noop,
};

Button.propTypes = {
  text: T.string.isRequired,
  onClick: T.func.isRequired,
  style: T.object,
};

export default Button;
