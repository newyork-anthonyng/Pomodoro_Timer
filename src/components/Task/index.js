import React from 'react';
import T from 'prop-types';
import noop from '../../utility/noop';
import glamorous from 'glamorous';

const H4 = glamorous.h4({
  display: 'inline-block',
  margin: 0,
});

const P = glamorous.p({
  marginTop: '10px',
});

const Button = glamorous.button({
  float: 'right',
});

const Task = ({ id, title, text, onDeleteClick }) => (
  <div data-taskId={id}>
    <H4>{title}</H4>

    <Button onClick={() => onDeleteClick(id)}>Ã—</Button>

    <P>{text}</P>
  </div>
);

Task.defaultProps = {
  onDeleteClick: noop,
};

Task.propTypes = {
  id: T.string.isRequired,
  title: T.string.isRequired,
  text: T.string.isRequired,
  onDeleteClick: T.func.isRequired,
};

export default Task;
