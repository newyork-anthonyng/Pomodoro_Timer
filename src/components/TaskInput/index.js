import React, { PureComponent } from 'react';
import T from 'prop-types';
import noop from '../../utility/noop';
import glamorous from 'glamorous';

const Container = glamorous.div({
  marginTop: '20px',
});
const Textarea = glamorous.textarea({
  display: 'block',
  fontSize: '16px',
  height: '100px',
  marginBottom: '10px',
  width: '340px',
});

const TaskInput = ({ value, onChange, onSubmit }) => (
  <Container>
    <Textarea
      value={value}
      onChange={onChange}
    />
    <button onClick={onSubmit}>âœ“</button>
  </Container>
);

TaskInput.defaultProps = {
  onSubmit: noop,
  onChange: noop,
};

TaskInput.propTypes = {
  value: T.string.isRequired,
  onSubmit: T.func.isRequired,
  onChange: T.func.isRequired,
};

export default TaskInput;
