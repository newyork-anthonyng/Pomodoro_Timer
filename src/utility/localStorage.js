const Storage = (function() {
  const KEY = 'pomodoro_timer';

  const get = () => {
    try {
      return JSON.parse(window.localStorage.getItem(KEY));
    } catch (e) {
      return [];
    }
  };

  const set = (favorites) => {
    window.localStorage.setItem(KEY, JSON.stringify(favorites));
  };

  return {
    get,
    set,
  };
})();

export default Storage;
