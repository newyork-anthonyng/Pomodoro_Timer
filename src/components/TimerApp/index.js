import React from 'react';
import T from 'prop-types';
import noop from '../../utility/noop';
import TimeDisplay from '../TimeDisplay';
import Button from '../Button';
import TaskInput from '../TaskInput';
import TaskList from '../TaskList';

const TimerApp = ({
  onPlayClick,
  onPauseClick,
  onResetClick,
  timeInMs,
  timerIsActive,
  tasks,
  onTaskInputChange,
  taskInputValue,
  onTaskSubmit,
  onTaskDeleteClick,
  onClearAllClick,
}) => (
  <div>
    <TimeDisplay timeInMs={timeInMs} />

    <Button
      text={timerIsActive ? 'Pause' : 'Play'}
      onClick={timerIsActive ? onPauseClick : onPlayClick}
      style={{ marginRight: '20px' }}
    />

    <Button
      text="Reset"
      onClick={onResetClick}
    />

    <TaskInput
      value={taskInputValue}
      onSubmit={onTaskSubmit}
      onChange={onTaskInputChange}
    />

    <TaskList
      tasks={tasks}
      onDeleteClick={onTaskDeleteClick}
    />

    <Button
      text="Clear All"
      onClick={onClearAllClick}
    />

  </div>
);

TimerApp.defaultProps = {
  onPlayClick: noop,
  onPauseClick: noop,
  onResetClick: noop,
  onTaskInputChange: noop,
  onTaskSubmit: noop,
  onTaskDeleteClick: noop,
  onClearAllClick: noop,
};

TimerApp.propTypes = {
  onPlayClick: T.func.isRequired,
  onPauseClick: T.func.isRequired,
  onResetClick: T.func.isRequired,
  timeInMs: T.number.isRequired,
  timerIsActive: T.bool.isRequired,
  tasks: T.arrayOf(T.shape({
    title: T.string.isRequired,
    text: T.string.isRequired,
  })).isRequired,
  onTaskInputChange: T.func.isRequired,
  taskInputValue: T.string.isRequired,
  onTaskSubmit: T.func.isRequired,
  onTaskDeleteClick: T.func.isRequired,
  onClearAllClick: T.func.isRequired,
};

export default TimerApp;
