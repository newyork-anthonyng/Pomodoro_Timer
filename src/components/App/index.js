import React, { Component } from 'react';
import TimerApp from '../TimerApp';
import TimerWorker from '../../utility/webworkerWrapper';
import WebNotifications from '../WebNotifications';
import Title from '../Title';
import uuid from '../../utility/uuid';
import timeFormatter from '../../utility/timeFormatter';
import localStorage from '../../utility/localStorage';

class App extends Component {
  constructor() {
    super();

    this.WORK_TIME = 25 * 60 * 1000; // 25 minutes
    this.BREAK_TIME = 5 * 60 * 1000; // 5 minutes

    this.state = {
      isWorking: true,
      timeInMs: this.WORK_TIME,
      timerIsActive: false,
      tasks: [],
      taskInputValue: '',
    };

    this.createWorker = this.createWorker.bind(this);
    this.fetchTasks = this.fetchTasks.bind(this);
    this.handlePlayClick = this.handlePlayClick.bind(this);
    this.handlePauseClick = this.handlePauseClick.bind(this);
    this.handleResetClick = this.handleResetClick.bind(this);
    this.handleTaskInputChange = this.handleTaskInputChange.bind(this);
    this.handleTaskSubmit = this.handleTaskSubmit.bind(this);
    this.handleDeleteClick = this.handleDeleteClick.bind(this);
  }

  componentDidMount() {
    this.createWorker();
    this.fetchTasks();
  }

  createWorker() {
    this.timerWorker = new TimerWorker();

    this.timerWorker.onmessage = (e) => {
      switch (e.data.action) {
      case 'TICK':
        this.setState({ timeInMs: e.data.time });
        break;
      case 'COMPLETE':
        this.setState({
          timerIsActive: false,
          isWorking: !this.state.isWorking,
          timeInMs: this.state.isWorking ? this.BREAK_TIME : this.WORK_TIME,
        }, () => {
          if (this.notification) {
            this.notification.createNotification({
              body: this.state.isWorking ? 'Back to work' : 'Take a break',
            });
          }
        });
        break;
      default:
        break;
      }
    };
  }

  fetchTasks() {
    this.setState({
      tasks: localStorage.get() || [],
    });
  }

  handlePlayClick() {
    this.timerWorker.postMessage({
      action: 'START',
      time: this.state.timeInMs,
    });

    this.setState({
      timerIsActive: true,
    });
  }

  handlePauseClick() {
    this.timerWorker.postMessage({
      action: 'STOP',
    });

    this.setState({
      timerIsActive: false,
    });
  }

  handleResetClick() {
    this.timerWorker.postMessage({
      action: 'STOP',
    });

    this.setState({
      timerIsActive: false,
      timeInMs: this.WORK_TIME,
      isWorking: true,
    });
  }

  handleTaskInputChange(e) {
    this.setState({
      taskInputValue: e.target.value,
    });
  }

  handleTaskSubmit() {
    if (this.state.taskInputValue.trim().length === 0) {
      return;
    }

    const newTasks = [{
      id: `${uuid()}`,
      title: `${this.state.tasks.length + 1}`,
      text: this.state.taskInputValue,
    }].concat(this.state.tasks);

    this.setState({
      tasks: newTasks,
      taskInputValue: '',
    });

    localStorage.set(newTasks);
  }

  handleDeleteClick(id) {
    const newTasks = this.state.tasks.filter((task) => task.id !== id);

    this.setState({
      tasks: newTasks,
    });

    localStorage.set(newTasks);
  }

  render() {
    const {
      timeInMs,
      timerIsActive,
      tasks,
      taskInputValue,
    } = this.state;

    return (
      <div>
        <Title text={timeFormatter(this.state.timeInMs)} />

        <TimerApp
          onPlayClick={this.handlePlayClick}
          onPauseClick={this.handlePauseClick}
          onResetClick={this.handleResetClick}
          timeInMs={timeInMs}
          timerIsActive={timerIsActive}
          tasks={tasks}
          onTaskInputChange={this.handleTaskInputChange}
          taskInputValue={taskInputValue}
          onTaskSubmit={this.handleTaskSubmit}
          onTaskDeleteClick={this.handleDeleteClick}
        />
        <WebNotifications ref={(notification) => this.notification = notification} />
      </div>
    );
  }
}

export default App;
