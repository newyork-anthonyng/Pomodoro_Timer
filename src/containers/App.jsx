import React from 'react';
import { connect } from 'react-redux';
import { TimeContainer } from './TimeContainer';
import { TimerLabels } from '../components/TimerLabels';
import { SettingsContainer } from './SettingsContainer';
import { Footer } from '../components/Footer';
import { GiphyContainer } from './GiphyContainer';
import { NotificationContainer } from './NotificationContainer';
import { startTimer, setTime, stopTimer, toggleMode, toggleSettingsPanel } from '../actions';
import Utility from '../util/utility';
const TimerWorker = require('worker!./webworker.js');

let App = React.createClass({
	componentDidMount: function() {
		this.setWorker();
		this.setKeyboardShortcut();
	},

	componentDidUpdate: function() {
		document.title = Utility.formatTime(this.props.seconds);
	},

	setWorker: function() {
		this.timerWorker = new TimerWorker();

		this.timerWorker.addEventListener('message', (e) => {
			switch(e.data.action) {
				case 'TICK':
					this.props.setTime(e.data.time);
					break;
				case 'COMPLETE':
					this.props.stopTimer();

					if(this.props.mode === 'work') {
						this.props.setTime(this.props.default.break);
						this.playTimer();
					} else {
						this.props.setTime(this.props.default.work);
					}

					this.props.toggleMode();
					break;
			}
		});
	},

	setKeyboardShortcut: function() {
		const SPACE_KEY = 32;
		const R_KEY = 82;
		const S_KEY = 83;

		window.addEventListener('keydown', (e) => {
			if(e.keyCode === SPACE_KEY) this.handlePlayClick();
			if(e.keyCode === R_KEY) this.resetTimer();
			if(e.keyCode === S_KEY) this.toggleSettings();
		});
	},

	handlePlayClick: function() {
		if(!this.props.isRunning) {
			this.playTimer();
		} else {
			this.pauseTimer();
		}
	},

	playTimer: function() {
		this.timerWorker.postMessage({
			action: 'START',
			time: this.props.seconds
		});

		this.props.startTimer();
	},

	pauseTimer: function() {
		this.timerWorker.postMessage({
			action: 'STOP'
		});

		this.props.stopTimer();
	},

	resetTimer: function() {
		this.pauseTimer();
		this.props.setTime(this.props.default.work);
	},

	toggleSettings: function() {
		this.props.toggleSettingsPanel();
	},

	render: function() {
		return (
			<div>
				<TimeContainer />
				<TimerLabels
					handlePlayClick={this.handlePlayClick}
					handleResetClick={this.resetTimer}
					handleSettingsClick={this.toggleSettings}
					isRunning={this.props.isRunning}
				/>
				<SettingsContainer />
				<Footer />
				<GiphyContainer mode={this.props.mode} />
				<NotificationContainer />
			</div>
		);
	}
});

const mapStateToProps = (state) => {
	return state;
};

const mapDispatchToProps = (dispatch) => {
	return {
		startTimer: () => {
			dispatch(startTimer());
		},
		setTime: (seconds) => {
			dispatch(setTime(seconds));
		},
		stopTimer: () => {
			dispatch(stopTimer());
		},
		toggleMode: () => {
			dispatch(toggleMode());
		},
		toggleSettingsPanel: () => {
			dispatch(toggleSettingsPanel());
		}
	};
};

export default connect(
	mapStateToProps,
	mapDispatchToProps
)(App);
