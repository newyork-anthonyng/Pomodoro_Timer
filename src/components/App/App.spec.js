import React from 'react';
import { mount } from 'enzyme';

jest.mock('../../utility/webworkerWrapper');
import TimerWorker from '../../utility/webworkerWrapper';

jest.mock('../WebNotifications');

jest.mock('../../utility/uuid');

jest.mock('../../utility/localStorage');
import localStorage from '../../utility/localStorage';

import TimerApp from '../TimerApp';
import TaskInput from '../TaskInput';
import Button from '../Button';
import Task from '../Task';
import App from './';

/*
* Shhh...remove console from jest output
*/
// eslint-disable-next-line no-console
console.error = jest.fn();
beforeEach(() => {
  localStorage.set = jest.fn();
  localStorage.get = jest.fn(() => []);
});

describe('On component mount', () => {
  it('should create new web worker on mount', () => {
    const wrapper = mount(<App />);

    expect(TimerWorker.mock.calls.length).toEqual(1);
    const webworker = wrapper.instance().timerWorker;
    expect(typeof webworker.onmessage).toEqual('function');
  });

  it('should fetch items from local storage on mount', () => {
    const tasks = [
      { id: 'apple1', title: 'Apple', text: 'Ate an apple' },
      { id: 'banana2', title: 'Banana', text: 'Ate a banana' },
    ];
    localStorage.get = jest.fn(() => tasks);
    const wrapper = mount(<App />);

    expect(localStorage.get).toHaveBeenCalledTimes(1);
    expect(wrapper.find(TimerApp).props().tasks).toEqual(tasks);
  });

  it('should return default empty array when local storage contains nothing', () => {
    localStorage.get = jest.fn(() => null);
    const wrapper = mount(<App />);

    expect(localStorage.get).toHaveBeenCalledTimes(1);
    expect(wrapper.find(TimerApp).props().tasks).toEqual([]);
  });
});

describe('When clicking play button', () => {
  it('should update timerIsActive to true', () => {
    const wrapper = mount(<App />);

    const playButton = wrapper.find('button').at(0);
    playButton.simulate('click');

    expect(wrapper.find(TimerApp).props().timerIsActive).toBeTruthy();
  });

  it('should start web worker', () => {
    const wrapper = mount(<App />);

    const playButton = wrapper.find('button').at(0);
    playButton.simulate('click');

    const webworker = wrapper.instance().timerWorker;
    expect(webworker.postMessage).toHaveBeenCalledTimes(1);
    expect(webworker.postMessage.mock.calls[0][0]).toMatchSnapshot();
  });
});

describe('When clicking pause button', () => {
  it('should update timerIsActive to false', () => {
    const wrapper = mount(<App />);
    wrapper.setState({ timerIsActive: true });

    const pauseButton = wrapper.find('button').at(0);
    pauseButton.simulate('click');

    expect(wrapper.find(TimerApp).props().timerIsActive).toBeFalsy();
  });

  it('should stop web worker', () => {
    const wrapper = mount(<App />);
    wrapper.setState({ timerIsActive: true });

    const pauseButton = wrapper.find('button').at(0);
    pauseButton.simulate('click');

    const webworker = wrapper.instance().timerWorker;
    expect(webworker.postMessage).toHaveBeenCalledTimes(1);
    expect(webworker.postMessage.mock.calls[0][0]).toMatchSnapshot();
  });
});

describe('When clicking reset button', () => {
  it('should update timerIsActive to false', () => {
    const wrapper = mount(<App />);
    wrapper.setState({ timerIsActive: true });

    const resetButton = wrapper.find('button').at(1);
    resetButton.simulate('click');

    expect(wrapper.find(TimerApp).props().timerIsActive).toBeFalsy();
  });

  it('should reset timer', () => {
    const wrapper = mount(<App />);
    wrapper.setState({ timeInMs: 42 });

    const resetButton = wrapper.find('button').at(1);
    resetButton.simulate('click');

    const workTime = wrapper.instance().WORK_TIME;
    expect(wrapper.find(TimerApp).props().timeInMs).toEqual(workTime);
  });

  it('should stop web worker', () => {
    const wrapper = mount(<App />);

    const resetButton = wrapper.find('button').at(1);
    resetButton.simulate('click');

    const webworker = wrapper.instance().timerWorker;
    expect(webworker.postMessage).toHaveBeenCalledTimes(1);
    expect(webworker.postMessage.mock.calls[0][0]).toMatchSnapshot();
  });

  it('should set mode to WORK', () => {
    const wrapper = mount(<App />);
    wrapper.setState({
      isWorking: false,
    });

    const resetButton = wrapper.find('button').at(1);
    resetButton.simulate('click');

    expect(wrapper.state().isWorking).toBeTruthy();
  });
});

describe('TimerWorker messages', () => {
  it('should update time on every TICK message', () => {
    const wrapper = mount(<App />);
    const webworker = wrapper.instance().timerWorker;

    const newTime = 42;
    webworker.onmessage({
      data: {
        action: 'TICK',
        time: newTime,
      },
    });

    expect(wrapper.find(TimerApp).props().timeInMs).toEqual(newTime);
  });

  it('should update timerIsActive to false on COMPLETE message', () => {
    const wrapper = mount(<App />);
    wrapper.setState({ timerIsActive: true });
    const webworker = wrapper.instance().timerWorker;

    webworker.onmessage({
      data: {
        action: 'COMPLETE',
      },
    });

    expect(wrapper.find(TimerApp).props().timerIsActive).toBeFalsy();
  });

  it('should update time when work pomodoro is completed', () => {
    const wrapper = mount(<App />);
    const webworker = wrapper.instance().timerWorker;

    webworker.onmessage({
      data: {
        action: 'COMPLETE',
      },
    });

    const breakTime = wrapper.instance().BREAK_TIME;
    expect(wrapper.find(TimerApp).props().timeInMs).toEqual(breakTime);
  });

  it('should update time when break pomodoro is completed', () => {
    const wrapper = mount(<App />);
    const webworker = wrapper.instance().timerWorker;
    wrapper.setState({ isWorking: false });

    webworker.onmessage({
      data: {
        action: 'COMPLETE',
      },
    });

    const workTime = wrapper.instance().WORK_TIME;
    expect(wrapper.find(TimerApp).props().timeInMs).toEqual(workTime);
  });

  it('should toggle from work to break on COMPLETE message', () => {
    const wrapper = mount(<App />);
    const webworker = wrapper.instance().timerWorker;
    wrapper.setState({ isWorking: true });

    webworker.onmessage({
      data: {
        action: 'COMPLETE',
      },
    });

    expect(wrapper.state().isWorking).toBeFalsy();
  });

  it('should toggle from break to work on COMPLETE message', () => {
    const wrapper = mount(<App />);
    const webworker = wrapper.instance().timerWorker;
    wrapper.setState({ isWorking: false });

    webworker.onmessage({
      data: {
        action: 'COMPLETE',
      },
    });

    expect(wrapper.state().isWorking).toBeTruthy();
  });
});

describe('WebNotifications', () => {
  it('should display correct notification when WORK timer is completed', () => {
    const wrapper = mount(<App />);
    const webworker = wrapper.instance().timerWorker;
    wrapper.setState({ isWorking: true });

    webworker.onmessage({
      data: {
        action: 'COMPLETE',
      },
    });

    const notification = wrapper.instance().notification;
    expect(notification.createNotification).toHaveBeenCalledTimes(1);
    expect(notification.createNotification.mock.calls[0][0]).toMatchSnapshot();
  });

  it('should display correct notification when BREAK timer is completed', () => {
    const wrapper = mount(<App />);
    const webworker = wrapper.instance().timerWorker;
    wrapper.setState({ isWorking: false });

    webworker.onmessage({
      data: {
        action: 'COMPLETE',
      },
    });

    const notification = wrapper.instance().notification;
    expect(notification.createNotification).toHaveBeenCalledTimes(1);
    expect(notification.createNotification.mock.calls[0][0]).toMatchSnapshot();
  });
});

describe('Task inputs', () => {
  it('should update task input text when typing', () => {
    const wrapper = mount(<App />);

    const newText = 'pomodoro is awesome';
    const textArea = wrapper.find('textarea');
    textArea.simulate('change', { target: { value: newText  } });

    expect(textArea.props().value).toEqual(newText);
  });

  it('should add task when clicking on button', () => {
    const wrapper = mount(<App />);

    const textArea = wrapper.find('textarea');
    textArea.simulate('change', { target: { value: 'pomodoro is awesome' } });
    const button = wrapper.find(TaskInput).find('button');
    button.simulate('click');

    expect(wrapper.find(TimerApp).props().tasks).toMatchSnapshot();
  });

  it('should not add an empty task', () => {
    const wrapper = mount(<App />);

    const textArea = wrapper.find('textarea');
    textArea.simulate('change', { target: { value: ' ' } });
    textArea.simulate('keypress', { key: 'Enter' });

    expect(wrapper.find(TimerApp).props().tasks).toHaveLength(0);
  });

  it('should empty text input field after adding a task', () => {
    const wrapper = mount(<App />);

    const textArea = wrapper.find('textarea');
    textArea.simulate('change', { target: { value: 'pomodoro is awesome' } });
    const button = wrapper.find(TaskInput).find('button');
    button.simulate('click');

    expect(textArea.props().value).toEqual('');
  });

  it('should set localStorage after adding a task', () => {
    localStorage.set = jest.fn();

    const wrapper = mount(<App />);

    const textArea = wrapper.find('textarea');
    textArea.simulate('change', { target: { value: 'pomodoro is awesome' } });
    const button = wrapper.find(TaskInput).find('button');
    button.simulate('click');

    expect(localStorage.set).toHaveBeenCalledTimes(1);
  });

  it('should add task in correct order', () => {
    const wrapper = mount(<App />);
    const tasks = [
      { id: 'apple1', title: 'Apples', text: 'Ate an apple' },
      { id: 'banana2', title: 'Bananas', text: 'Ate a banana' },
    ];
    wrapper.setState({ tasks });

    const textArea = wrapper.find('textarea');
    textArea.simulate('change', { target: { value: 'pomodoro is awesome' } });
    const button = wrapper.find(TaskInput).find('button');
    button.simulate('click');

    expect(wrapper.find(TimerApp).props().tasks).toMatchSnapshot();
  });
});

describe('Task delete', () => {
  const tasks = [
    { id: 'apple1', title: 'Apples', text: 'Ate an apple' },
    { id: 'banana2', title: 'Bananas', text: 'Ate a banana' },
  ];

  it('should delete task', () => {
    const wrapper = mount(<App />);
    wrapper.setState({ tasks });

    const bananaTask = wrapper.find(Task).at(1);
    const deleteButton = bananaTask.find('button');
    deleteButton.simulate('click');

    expect(wrapper.find(TimerApp).props().tasks).toHaveLength(1);
  });

  it('should update the local storage', () => {
    const wrapper = mount(<App />);
    wrapper.setState({ tasks });
    localStorage.set.mockClear();

    const bananaTask = wrapper.find(Task).at(1);
    const deleteButton = bananaTask.find('button');
    deleteButton.simulate('click');

    expect(localStorage.set).toHaveBeenCalledTimes(1);
    expect(localStorage.set.mock.calls[0][0]).toMatchSnapshot();
  });
});

describe('Clear all tasks', () => {
  const tasks = [
    { id: 'apple1', title: 'Apples', text: 'Ate an apple' },
    { id: 'banana2', title: 'Bananas', text: 'Ate a banana' },
  ];

  it('should delete all tasks', () => {
    const wrapper = mount(<App />);
    wrapper.setState({ tasks });

    const clearAllButton = wrapper.find(Button).at(2);
    clearAllButton.simulate('click');

    expect(wrapper.find(TimerApp).props().tasks).toEqual([]);
  });

  it('should update the local storage', () => {
    const wrapper = mount(<App />);

    const clearAllButton = wrapper.find(Button).at(2);
    clearAllButton.simulate('click');

    expect(localStorage.set).toHaveBeenCalledTimes(1);
    expect(localStorage.set.mock.calls[0][0]).toMatchSnapshot();
  });
});

describe('Keyboard shortcuts', () => {
  let map;

  beforeEach(() => {
    map = {};

    document.addEventListener = jest.fn((event, cb) => {
      map[event] = cb;
    });
  });

  it('should add keypress event listener on mount', () => {
    mount(<App />);

    expect(document.addEventListener).toHaveBeenCalledTimes(1);
    expect(map.keypress).toBeDefined();
  });

  it('should PLAY when pressing ctrl + q and timer is not active', () => {
    const wrapper = mount(<App />);
    wrapper.setState({ timerIsActive: false });
    wrapper.instance().handlePlayClick = jest.fn();

    map.keypress({
      key: 'q',
      ctrlKey: true,
    });

    expect(wrapper.instance().handlePlayClick).toHaveBeenCalledTimes(1);
  });

  it('should PAUSE when pressing ctrl + q and timer is active', () => {
    const wrapper = mount(<App />);
    wrapper.setState({ timerIsActive: true });
    wrapper.instance().handlePauseClick = jest.fn();

    map.keypress({
      key: 'q',
      ctrlKey: true,
    });

    expect(wrapper.instance().handlePauseClick).toHaveBeenCalledTimes(1);
  });

  it('should RESET when pressing ctrl + R', () => {
    const wrapper = mount(<App />);
    wrapper.instance().handleResetClick = jest.fn();

    map.keypress({
      key: 'r',
      ctrlKey: true,
    });

    expect(wrapper.instance().handleResetClick).toHaveBeenCalledTimes(1);
  });

  it('should submit task when pressing ctrl + ENTER', () => {
    const wrapper = mount(<App />);
    wrapper.instance().handleTaskSubmit = jest.fn();

    map.keypress({
      key: 'Enter',
      ctrlKey: true,
    });

    expect(wrapper.instance().handleTaskSubmit).toHaveBeenCalledTimes(1);
  });

  it('should do nothing if ctrl key is not pressed', () => {
    const wrapper = mount(<App />);
    wrapper.instance().handlePauseClick = jest.fn();

    map.keypress({ key: 'q' });

    expect(wrapper.instance().handlePauseClick).not.toHaveBeenCalled();
  });
});
