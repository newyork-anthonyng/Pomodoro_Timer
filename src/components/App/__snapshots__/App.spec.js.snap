// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clear all tasks should update the local storage 1`] = `Array []`;

exports[`Task delete should update the local storage 1`] = `
Array [
  Object {
    "id": "apple1",
    "text": "Ate an apple",
    "title": "Apples",
  },
]
`;

exports[`Task inputs should add task in correct order 1`] = `
Array [
  Object {
    "id": "42",
    "text": "pomodoro is awesome",
    "title": "3",
  },
  Object {
    "id": "apple1",
    "text": "Ate an apple",
    "title": "Apples",
  },
  Object {
    "id": "banana2",
    "text": "Ate a banana",
    "title": "Bananas",
  },
]
`;

exports[`Task inputs should add task when clicking on button 1`] = `
Array [
  Object {
    "id": "42",
    "text": "pomodoro is awesome",
    "title": "1",
  },
]
`;

exports[`WebNotifications should display correct notification when BREAK timer is completed 1`] = `
Object {
  "body": "Back to work",
}
`;

exports[`WebNotifications should display correct notification when WORK timer is completed 1`] = `
Object {
  "body": "Take a break",
}
`;

exports[`When clicking pause button should stop web worker 1`] = `
Object {
  "action": "STOP",
}
`;

exports[`When clicking play button should start web worker 1`] = `
Object {
  "action": "START",
  "time": 1500000,
}
`;

exports[`When clicking reset button should stop web worker 1`] = `
Object {
  "action": "STOP",
}
`;
